### ISSUE TITLE
Disparate performance in diff methods

### ISSUE THREAD

**Issue by arlk** [JuliaArrays/StaticArrays.jl#541]
First of all, thank you for this great package! I'm a relatively new Julia user and I was using Static Arrays for some polynomial math that needs to use `diff`. I noticed that the benchmark speeds are very different for the two `diff` methods. MWE:
```julia
julia> A = @SMatrix randn(3,20);

julia> @btime diff($A,dims=$2);
  3.299 μs (2 allocations: 1.06 KiB)

julia> @btime diff($A,dims=$(Val(2)));
  37.283 ns (0 allocations: 0 bytes)
```

**Comment by andyferris** [comment]
Yes, this is known... but that's a larger discrepency then I expected.

When constant propagation works a bit better, we should be able to make the first method fast and deprecate the second.

**Comment by c42f** [comment]
Looking at the generated code it looks like our explicit `@inline` is being ignored for both these methods. Is that expected?

**Comment by c42f** [comment]
Probable dup of #540 ?
